---
title: "Find differentially expressed genes using edgeR library functions and reference genome assembly SL4 from September 2019"
author: "Ann Loraine and Molly Davis"
date: "`r Sys.Date()`"
output:
  pdf_document: default
  #html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Prelude - Define variables

```{r variables, message=FALSE, warning=FALSE}
genome_version="SL4"
counts_fname = paste0("results/muday-144-",genome_version,"_counts-salmon.txt")
sample_sheet_fname = "Documentation/muday-144_sample_sheet.xlsx"
all = NULL
library(git2r)
r = revparse_single(counts_fname,"HEAD")
hash = sha(r)
method="edgeR"
out_fname = paste0("results/CvT-",method,"-",genome_version,".txt")
```

* * *

# Introduction

This Markdown analyzes RNA-Seq gene expression data from data file 
``r counts_fname`` documented in ``r sample_sheet_fname`` using   differential expression analysis library `r method`.

The input data file was generated by nf-core/rna-seq pipeline that aligned RNA-Seq data and produced RNA-Seq fragment counts for  gene annotations from the release of the tomato genome known as `r genome_version`
released in September of 2019. 

In this Markdown, we aim to answer:

* How many genes were differentially expressed in treatment versus control comparisons?

# Input data file summary

The git hash (version) of ``r counts_fname`` was: 

* ``r hash``

The data file contains heat-treated and non-heat-treated samples from four tomato genotypes: `are` (anthocyanin reduced mutant), VF36 (wild-type tomato cultivar), and a VF36 line designated `F3H` containing a transgene encoding the `are` wild-type gene. Samples experienced the heat stress over a time course which included four time points: 15 minutes, 30 minutes, 45 minutes, and 75 minutes. Two temperatures were tested: 28 degrees C, the control, and 34 degrees C, the heat treatment.  

Samples and sample groups are named as follows:

* [genotype].[temperature].[treatment duration].[replicate number]

where genotype is A, V, or F; temperature is 28 or 34; and treatment duration is 15, 30, 25, or 75. 
There were three replicates per treatment group. 

In addition to answering the above question, we also aim to create a data file that we will use to compare differential expression results obtained here with differential expression results obtained using a different, but similar, R library called "DESeq2." If the results are similar, we can be more confident that we are using these libraries correctly. 

* * *

# Results

Load custom functions:

```{r message=FALSE, warning=FALSE}
source("Common.R")
source("StatFunctions.R")
```

Load required counts data:

```{r message=FALSE, warning=FALSE, paged.print=FALSE}
counts=getCounts(counts_fname,keep_description = T)
```

The table of RNA-Seq counts per gene loaded in the previous code chunk from file `r counts_fname` contained `r prettyNumber(nrow(counts))` rows corresponding to measured genes.


# Compute and tabulate differential expression

Define false discovery rate threshold for deciding whether a gene is differentially expressed:

```{r}
Q=0.05
```

We will use `r Q`, defined in the previous code chunk, to "call" a gene as differentially expressed.

As we test whether the treatment changed gene expression for each genotype and treatment duration combination, we will save results to a single data frame. At the end of the Markdown, we'll write this very large table to a file named ``r out_fname``.

## Genotype *are*, anthocyanin-reduced mutant

```{r}
group1_name = "A.28.15"
group2_name = "A.34.15"
results_table = getDeGenes(counts,group1_name,group2_name,method=method)
d=results_table
#d$PValue= ifelse(-log10(d$PValue) < 20,d$PValue,1e-20)
title=paste(group1_name,"vs",group2_name)

colorBlindGrey8   <- c("#999999", "#E69F00", "#56B4E9", "#009E73", 
                         "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
  if (method=="DESeq2") {
    lfc_colname = "log2FoldChange"
    padj_colname = "padj"
    p_colname = "pvalue"
    d$pvalue= ifelse(-log10(d$pvalue) < 20,d$pvalue,1e-20)
  }
  if (method=="edgeR") {
    lfc_colname = "logFC"
    padj_colname = "Q"
    p_colname = "PValue"
    d$PValue= ifelse(-log10(d$PValue) < 20,d$PValue,1e-20)
  }
  pCutoffCol = padj_colname
  if (!any(d[,pCutoffCol]<=Q)) {
    Q = round(min(d$padj),2)+0.01
  }
  lfcmin = min(d[,lfc_colname])
  lfcmax = max(d[,lfc_colname])
  #x_max = -log10(min(d[,p_colname]))
  keyvals <- ifelse(
    d[,lfc_colname] < -1, '#56B4E9',
    ifelse(d[,lfc_colname] > 1, '#CC79A7','black'))
  names(keyvals)[keyvals == '#CC79A7'] = 'Up-regulated'
  names(keyvals)[keyvals == 'black'] = 'Mid-regulated'
  names(keyvals)[keyvals == '#56B4E9'] = 'Down-regulated'
  
  v = EnhancedVolcano(d,
                      lab = d$gene,
                      x = lfc_colname,
                      y = p_colname,
                      selectLab = d[which(names(keyvals) %in% c('Up-regulated', 'Down-regulated')),
                                    "gene"],
                      colCustom = keyvals,
                      colAlpha = 1,
                      shape = c(4, 1, 6, 3),
                      pCutoff = Q,  
                      FCcutoff = 1.0,
                      pCutoffCol = pCutoffCol, # 
                      gridlines.minor=FALSE, 
                      gridlines.major=FALSE,
                      col = colorBlindGrey8,
                      #xlim = c(-10, 10),
                      #ylim = c(0,pmax),
                      #ylab = "-log(p)",
                      title = title, 
                      subtitle = paste("Horizontal dashed line shows FDR",Q),
                      legendPosition = "right")

v




volcano_A1 <- volcano_plot(results_table, paste(group1_name,"vs",group2_name),
                           method=method)
all = results_table
```

Comparing group `r group1_name` to `r group2_name` found  `r sum(results_table$Q<=Q)` genes with changed expression levels with false discovery rate Q less than or equal to `r Q`.


```{r}
group1_name = "A.28.30"
group2_name = "A.34.30"
results_table = getDeGenes(counts,group1_name,group2_name,method=method)
volcano_A2 <- volcano_plot(results_table, paste(group1_name,"vs",group2_name),
                           method=method)
all = rbind(all,results_table)
```

Comparing group `r group1_name` to `r group2_name` found  `r sum(results_table$Q<=Q)` genes with changed expression levels with false discovery rate Q less than or equal to `r Q`.

```{r}
group1_name = "A.28.45"
group2_name = "A.34.45"
results_table = getDeGenes(counts,group1_name,group2_name,method=method)
volcano_A3 <- volcano_plot(results_table, paste(group1_name,"vs",group2_name),
                           method=method)
all = rbind(all,results_table)
```

Comparing group `r group1_name` to `r group2_name` found  `r sum(results_table$Q<=Q)` genes with changed expression levels with false discovery rate Q less than or equal to `r Q`.

```{r}
group1_name = "A.28.75"
group2_name = "A.34.75"
results_table = getDeGenes(counts,group1_name,group2_name,method=method)
volcano_A4 <- volcano_plot(results_table, paste(group1_name,"vs",group2_name),
                           method=method)
all = rbind(all,results_table)
```

Comparing group `r group1_name` to `r group2_name` found  `r sum(results_table$Q<=Q)` genes with changed expression levels with false discovery rate Q less than or equal to `r Q`.

The number of DE genes called at Q less than `r Q`increased
with treatment duration time.

Display volcano plots that summarize the above results:

```{r echo=TRUE, fig.height=18, fig.width=18}
A_combined<- ggpubr::ggarrange(volcano_A1, volcano_A2, volcano_A3, volcano_A4, # list of plots
                  labels = "AUTO",
                  font.label = list(size = 30), 
                  common.legend = T, # COMMON LEGEND
                  legend = "right", # legend position
                  align = "hv", # Align them both, horizontal and vertical
                  nrow = 2, 
                  ncol = 2)
A_combined
```

The above plot confirms that the number of genes found to differentially expressed
within a time point increased with treatment duration.  

Also, most of the genes that changed were higher in the treatment versus control.


## Genotype *VF36*, wild-type 


```{r}
group1_name = "V.28.15"
group2_name = "V.34.15"
results_table = getDeGenes(counts,group1_name,group2_name,method=method)
volcano_A1 <- volcano_plot(results_table, paste(group1_name,"vs",group2_name),
                           method=method)
all = rbind(all,results_table)
```

Comparing group `r group1_name` to `r group2_name` found  `r sum(results_table$Q<=Q)` genes with changed expression levels with false discovery rate Q less than or equal to `r Q`.

```{r}
group1_name = "V.28.30"
group2_name = "V.34.30"
results_table = getDeGenes(counts,group1_name,group2_name,method=method)
volcano_A2 <- volcano_plot(results_table, paste(group1_name,"vs",group2_name),
                           method=method)
all = rbind(all,results_table)
```

Comparing group `r group1_name` to `r group2_name` found  `r sum(results_table$Q<=Q)` genes with changed expression levels with false discovery rate Q less than or equal to `r Q`.

```{r}
group1_name = "V.28.45"
group2_name = "V.34.45"
results_table = getDeGenes(counts,group1_name,group2_name,method=method)
volcano_A3 <- volcano_plot(results_table, paste(group1_name,"vs",group2_name),
                           method=method)
all = rbind(all,results_table)
```

Comparing group `r group1_name` to `r group2_name` found  `r sum(results_table$Q<=Q)` genes with changed expression levels with false discovery rate Q less than or equal to `r Q`.

```{r}
group1_name = "V.28.75"
group2_name = "V.34.75"
results_table = getDeGenes(counts,group1_name,group2_name,method=method)
volcano_A4 <- volcano_plot(results_table, paste(group1_name,"vs",group2_name),
                           method=method)
all = rbind(all,results_table)
```

Comparing group `r group1_name` to `r group2_name` found  `r sum(results_table$Q<=Q)` genes with changed expression levels with false discovery rate Q less than or equal to `r Q`.

The number of DE genes called at Q less than or equal to `r Q` increased
with treatment duration time.

Display volcano plots that summarize the above results:

```{r echo=TRUE, fig.height=18, fig.width=18}
A_combined<- ggpubr::ggarrange(volcano_A1, volcano_A2, volcano_A3, volcano_A4, # list of plots
                  labels = "AUTO",
                  font.label = list(size = 30), 
                  common.legend = T, # COMMON LEGEND
                  legend = "right", # legend position
                  align = "hv", # Align them both, horizontal and vertical
                  nrow = 2, 
                  ncol = 2)
A_combined
```


As with the `are` genotype, the above plot confirms that the number of genes found to differentially expressed
within a time point increased with treatment duration.

Also, most of the genes that changed were higher in the treatment versus control.


## Genotype *F3H-OX3*, F3H overexpression genotype

```{r}
group1_name = "F.28.15"
group2_name = "F.34.15"
results_table = getDeGenes(counts,group1_name,group2_name,method=method)
volcano_A1 <- volcano_plot(results_table, paste(group1_name,"vs",group2_name),
                           method=method)
all = rbind(all,results_table)
```

Comparing group `r group1_name` to `r group2_name` found  `r sum(results_table$Q<=Q)` genes with changed expression levels with false discovery rate Q less than or equal to `r Q`.

```{r}
group1_name = "F.28.30"
group2_name = "F.34.30"
results_table = getDeGenes(counts,group1_name,group2_name,method=method)
volcano_A2 <- volcano_plot(results_table, paste(group1_name,"vs",group2_name),
                           method=method)
all = rbind(all,results_table)
```

Comparing group `r group1_name` to `r group2_name` found  `r sum(results_table$Q<=Q)` genes with changed expression levels with false discovery rate Q less than or equal to `r Q`.

```{r}
group1_name = "F.28.45"
group2_name = "F.34.45"
results_table = getDeGenes(counts,group1_name,group2_name,method=method)
volcano_A3 <- volcano_plot(results_table, paste(group1_name,"vs",group2_name),
                           method=method)
all = rbind(all,results_table)
```

Comparing group `r group1_name` to `r group2_name` found  `r sum(results_table$Q<=Q)` genes with changed expression levels with false discovery rate Q less than or equal to `r Q`.

```{r}
group1_name = "F.28.75"
group2_name = "F.34.75"
results_table = getDeGenes(counts,group1_name,group2_name,method=method)
volcano_A4 <- volcano_plot(results_table, paste(group1_name,"vs",group2_name),
                           method=method)
all = rbind(all,results_table)
```

Comparing group `r group1_name` to `r group2_name` found  `r sum(results_table$Q<=Q)` genes with changed expression levels with false discovery rate Q less than or equal to `r Q`.

The number of DE genes called at Q less than or equal to `r Q` increased
with treatment duration time.

Display volcano plots that summarize the above results_table:

```{r echo=TRUE, fig.height=18, fig.width=18}
A_combined<- ggpubr::ggarrange(volcano_A1, volcano_A2, volcano_A3, volcano_A4, # list of plots
                  labels = "AUTO",
                  font.label = list(size = 30), 
                  common.legend = T, # COMMON LEGEND
                  legend = "right", # legend position
                  align = "hv", # Align them both, horizontal and vertical
                  nrow = 2, 
                  ncol = 2)
A_combined
```

As with the `are` genotype, the above plot confirms that the number of genes found to differentially expressed
within a time point increased with treatment duration.

Also, most of the genes that changed were higher in the treatment versus control.

# Write results to a file for analysts' convenience

All DE results were saved to a large data frame, saved to data frame `all`, with `r prettyNum(nrow(all))` rows and `r ncol(all)` columns.

Format and organize results:

```{r}
all$logFC=round(all$logFC,3)
all$logCPM=round(all$logCPM,3)
all$Q=signif(all$Q,3)
all$PValue=signif(all$PValue,3)
all = all[,c("gene","group1","group2","Q","PValue","logFC",
             "logCPM","description")]
```

Write all the results to a data file:

```{r}
write.table(all,file=out_fname,quote=F,row.names = F,sep="\t")
```

A file was created named `r out_fname` that contains all the results.

Explanation of columns:

* gene - `r genome_version` gene measured
* group 1 - control group
* group 2 - treatment group
* Q - false discovery rate; adjusted p-value computed using method of Benjamini and Hochberg
* PValue - nominal (unadjusted) p value 
* logFC - log2(group 2 average/group 1 average)
* logCPM - ?
* description - gene description 

* * *

# Discussion

Within each genotype, the number of genes exhibiting expression changes increased with treatment duration. 

Note that the different genotypes exhibited different numbers of differentially expressed genes, with the `are` genotype exhibiting the greatest number. This is consistent with the known `are` lower fertility mutant phenotype. 

However, comparing numbers of differentially expressed across comparisons is not the best way to assess the influence of genotype on changes in gene expression detected between treatment and control samples. A better, more rigorous approach is to *simultaneously* test the interaction between genotype and treatment, by testing an interaction term in a linear model. We will do this in a different Markdown document.

* * *

# Conclusion

* The number of genes detected as changed increased with treatment duration.

* The data suggest, but do not prove, that heat stress triggers greater changes in gene expression for the `are` genotype than for the other two genotypes tested. 

* * *

# Session info

```{r}
sessionInfo()
```
