---
title: "Find differentially expressed genes using DESeq2 library functions and reference genome assembly SL5 from June 2022"
author: "Molly Davis and Ann Loraine"
date: "`r Sys.Date()`"
output:
  pdf_document: default
  # html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r variables, message=FALSE, warning=FALSE, include=FALSE}
counts_fname = "results/muday-144-SL5_counts-salmon.txt"
sample_sheet_fname = "Documentation/muday-144_sample_sheet.xlsx"
all = NULL
library(git2r)
r = revparse_single(counts_fname,"HEAD")
hash = sha(r)
method="DESeq2"
assembly = "SL5"
out_fname=paste0("results/CvT-",method,"-",assembly,".txt")
```

* * *

# Introduction

This Markdown analyzes RNA-Seq gene expression data from data file 
``r counts_fname`` documented in ``r sample_sheet_fname`` using   differential expression analysis library `r method`.

The input data file was generated by nf-core/rna-seq pipeline that aligned RNA-Seq data and produced RNA-Seq fragment counts for  gene annotations from the June 2022 release of the tomato genome 
assembly `r assembly` reported in open access article [Graph pangenome captures missing heritability and empowers tomato breeding](https://www.nature.com/articles/s41586-022-04808-9). 

Also, see the Sol Genomics Web page [Tomato graph pangenome project](https://solgenomics.net/projects/tgg).

In this Markdown, we aim to answer:

* How many genes were differentially expressed in treatment versus control comparisons?

# Input data file summary:

The git hash (version) of ``r counts_fname`` was: ``r hash``.

The data file contains heat-treated and non-heat-treated samples from four tomato genotypes: `are` (anthocyanin reduced mutant), VF36 (wild-type tomoato cultivar), and a VF36 line designated `F3H` containing a transgene encoding the `are` wild-type gene. Samples experienced the heat stress over a time course which included four time points. 

In addition to answering the above question, we also aim to create a data file that we will use to compare differential expression results obtained here with differential expression results obtained using a different, but similar, R library called "edgeR." If the results are similar, we can be more confident that we are using these libraries correctly. 

* * *

# Analysis and Results

Load custom functions for analyzing and visualizing differential expression:

```{r message=FALSE, warning=FALSE}
source("Common.R")
source("StatFunctions.R")
```

Load required counts data:

```{r message=FALSE, warning=FALSE, paged.print=FALSE}
counts=getCounts(counts_fname,keep_description = T)
```

The table of RNA-Seq counts per gene loaded in the previous code chunk from file `r counts_fname` contained `r prettyNumber(nrow(counts))` rows corresponding to measured genes.

Define false discovery rate threshold for deciding whether a gene is differentially expressed:

```{r}
Q=0.05
```

We will use `r Q`, defined in the previous code chunk, to "call" a gene as  differentially expressed.

As we test whether the treatment changed gene expression for each genotype and treatment duration combination, we will save results to a single data frame. At the end of the Markdown, we'll write this very large table to a file named ``r out_fname``.

# Genotype *are*, anthocyanin-reduced mutant

```{r message=FALSE, warning=FALSE}
group1_name = "A.28.15"
group2_name = "A.34.15"
results_table = getDeGenes(counts,group1_name,group2_name)
volcano_A1 <- volcano_plot(results_table, paste(group1_name,"vs",group2_name),
                           method=method)
all = results_table
```

Comparing group `r group1_name` to `r group2_name` found  `r sum(results_table$padj<=Q)` genes with changed expression levels with false discovery rate Q less than or equal to `r Q`.


```{r message=FALSE, warning=FALSE}
group1_name = "A.28.30"
group2_name = "A.34.30"
results_table = getDeGenes(counts,group1_name,group2_name)
volcano_A2 <- volcano_plot(results_table, paste(group1_name,"vs",group2_name),
                           method=method)
all = rbind(all,results_table)
```

Comparing group `r group1_name` to `r group2_name` found  `r sum(results_table$padj<=Q)` genes with changed expression levels with false discovery rate Q less than or equal to `r Q`.

```{r message=FALSE, warning=FALSE}
group1_name = "A.28.45"
group2_name = "A.34.45"
results_table = getDeGenes(counts,group1_name,group2_name)
volcano_A3 <- volcano_plot(results_table, paste(group1_name,"vs",group2_name),
                           method=method)
all = rbind(all,results_table)
```

Comparing group `r group1_name` to `r group2_name` found  `r sum(results_table$padj<=Q)` genes with changed expression levels with false discovery rate Q less than or equal to `r Q`.

```{r message=FALSE, warning=FALSE}
group1_name = "A.28.75"
group2_name = "A.34.75"
results_table = getDeGenes(counts,group1_name,group2_name)
volcano_A4 <- volcano_plot(results_table, paste(group1_name,"vs",group2_name),
                           method=method)
all = rbind(all,results_table)
```

Comparing group `r group1_name` to `r group2_name` found  `r sum(results_table$padj<=Q)` genes with changed expression levels with false discovery rate Q less than or equal to `r Q`.

The number of DE genes called at Q less than `r Q`increased
with treatment duration time.

Display volcano plots that summarize the above results:

```{r echo=TRUE, fig.height=18, fig.width=18}
A_combined<- ggpubr::ggarrange(volcano_A1, volcano_A2, volcano_A3, volcano_A4, # list of plots
                  labels = "AUTO",
                  font.label = list(size = 30), 
                  common.legend = T, # COMMON LEGEND
                  legend = "right", # legend position
                  align = "hv", # Align them both, horizontal and vertical
                  nrow = 2, 
                  ncol = 2)
A_combined
```

The volcano plots confirm the previous observation that the number of genes found
to be differentially expressed in response to the heat treatment increased with
treatment duration.

They also show that most of the changes were in the positive direction. Most of the
genes that met the significance threshold (the horizontal line) were in the upper
right quadrant of the plots, indicating that their expression levels increased
in response to the treatment.

# Genotype *VF36*, wild-type 


```{r message=FALSE, warning=FALSE}
group1_name = "V.28.15"
group2_name = "V.34.15"
results_table = getDeGenes(counts,group1_name,group2_name)
volcano_A1 <- volcano_plot(results_table, paste(group1_name,"vs",group2_name),
                           method=method)
all = rbind(all,results_table)
```

Comparing group `r group1_name` to `r group2_name` found  `r sum(results_table$padj<=Q)` genes with changed expression levels with false discovery rate Q less than or equal to `r Q`.

```{r message=FALSE, warning=FALSE}
group1_name = "V.28.30"
group2_name = "V.34.30"
results_table = getDeGenes(counts,group1_name,group2_name)
volcano_A2 <- volcano_plot(results_table, paste(group1_name,"vs",group2_name),
                           method=method)
all = rbind(all,results_table)
```

Comparing group `r group1_name` to `r group2_name` found  `r sum(results_table$padj<=Q)` genes with changed expression levels with false discovery rate Q less than or equal to `r Q`.

```{r message=FALSE, warning=FALSE}
group1_name = "V.28.45"
group2_name = "V.34.45"
results_table = getDeGenes(counts,group1_name,group2_name)
volcano_A3 <- volcano_plot(results_table, paste(group1_name,"vs",group2_name),
                           method=method)
all = rbind(all,results_table)
```

Comparing group `r group1_name` to `r group2_name` found  `r sum(results_table$padj<=Q)` genes with changed expression levels with false discovery rate Q less than or equal to `r Q`.

```{r message=FALSE, warning=FALSE}
group1_name = "V.28.75"
group2_name = "V.34.75"
results_table = getDeGenes(counts,group1_name,group2_name)
volcano_A4 <- volcano_plot(results_table, paste(group1_name,"vs",group2_name),
                           method=method)
all = rbind(all,results_table)
```

Comparing group `r group1_name` to `r group2_name` found  `r sum(results_table$padj<=Q)` genes with changed expression levels with false discovery rate Q less than or equal to `r Q`.

The number of DE genes called at Q less than or equal to `r Q` increased
with treatment duration time.

Display volcano plots that summarize the above results:

```{r echo=TRUE, fig.height=18, fig.width=18}
A_combined<- ggpubr::ggarrange(volcano_A1, volcano_A2, volcano_A3, volcano_A4, # list of plots
                  labels = "AUTO",
                  font.label = list(size = 30), 
                  common.legend = T, # COMMON LEGEND
                  legend = "right", # legend position
                  align = "hv", # Align them both, horizontal and vertical
                  nrow = 2, 
                  ncol = 2)
A_combined
```

The volcano plots recapitulate the previous observation that the number of genes found
to be differentially expressed in response to the heat treatment increased with
treatment duration.

Second, nearly all changes that did meet achieve the target false discovery rate of `r Q`
were changes in the positive direction with the average estimated treatment expression 
higher than control. This can be seen by observing how most of the points above
the horizontal line, when it was drawn, were also the in the right half of the
plot. 

Note that plot A draws at line at a different FDR, a value larger than `r Q`.
If no gene achieved FDR of `r Q` or higher, then our code picks a new Q indicating
the p value for the gene with the smallest adjusted p value observed. 

# Genotype *F3H-OX3*, F3H overexpression genotype

```{r message=FALSE, warning=FALSE}
group1_name = "F.28.15"
group2_name = "F.34.15"
results_table = getDeGenes(counts,group1_name,group2_name)
volcano_A1 <- volcano_plot(results_table, paste(group1_name,"vs",group2_name),
                           method=method)
all = rbind(all,results_table)
```

Comparing group `r group1_name` to `r group2_name` found  `r sum(results_table$padj<=Q)` genes with changed expression levels with false discovery rate Q less than or equal to `r Q`.

```{r message=FALSE, warning=FALSE}
group1_name = "F.28.30"
group2_name = "F.34.30"
results_table = getDeGenes(counts,group1_name,group2_name)
volcano_A2 <- volcano_plot(results_table, paste(group1_name,"vs",group2_name),
                           method=method)
all = rbind(all,results_table)
```

Comparing group `r group1_name` to `r group2_name` found  `r sum(results_table$padj<=Q)` genes with changed expression levels with false discovery rate Q less than or equal to `r Q`.

```{r message=FALSE, warning=FALSE}
group1_name = "F.28.45"
group2_name = "F.34.45"
results_table = getDeGenes(counts,group1_name,group2_name)
volcano_A3 <- volcano_plot(results_table, paste(group1_name,"vs",group2_name),
                           method=method)
all = rbind(all,results_table)
```

Comparing group `r group1_name` to `r group2_name` found  `r sum(results_table$padj<=Q)` genes with changed expression levels with false discovery rate Q less than or equal to `r Q`.

```{r message=FALSE, warning=FALSE}
group1_name = "F.28.75"
group2_name = "F.34.75"
results_table = getDeGenes(counts,group1_name,group2_name)
volcano_A4 <- volcano_plot(results_table, paste(group1_name,"vs",group2_name),
                           method=method)
all = rbind(all,results_table)
```

Comparing group `r group1_name` to `r group2_name` found  `r sum(results_table$padj<=Q)` genes with changed expression levels with false discovery rate Q less than or equal to `r Q`.

The number of DE genes called at Q less than or equal to `r Q` increased
with treatment duration time.

Display volcano plots that summarize the above results:

```{r echo=TRUE, fig.height=18, fig.width=18}
A_combined<- ggpubr::ggarrange(volcano_A1, volcano_A2, volcano_A3, volcano_A4, # list of plots
                  labels = "AUTO",
                  font.label = list(size = 30), 
                  common.legend = T, # COMMON LEGEND
                  legend = "right", # legend position
                  align = "hv", # Align them both, horizontal and vertical
                  nrow = 2, 
                  ncol = 2)
A_combined
```

The volcano plots recapitulate the previous observations that the number of genes found
to be differentially expressed in response to the heat treatment increased with
treatment duration.

Second, nearly all changes that did meet achieve the target false discovery rate of `r Q`
were changes in the positive direction with the average estimated treatment expression 
higher than control. This can be seen by observing how most of the points above
the horizontal line were also the in the right half of the plot. 

# Write results to a file for analysts' convenience

All DE results were saved to a large data frame, saved to data frame `all`, with `r prettyNum(nrow(all))` rows and `r ncol(all)` columns.

Organize results and round numeric results to 3 significant digits:

```{r}
all = all[,c("gene","group1","group2","baseMean","padj","pvalue","log2FoldChange","lfcSE","stat","description")]
for (i in 4:9) {
  all[,i]=signif(all[,i],3)
}
```

Add SL4 gene name as a new column:

```{r}
SL4_gene_names = getSL4GeneNames(all$description)
all$SL4 = SL4_gene_names
```

Write the data file:

```{r}
write.table(all,file=out_fname,quote=F,row.names = F,sep="\t")
```

A file was created named `r out_fname` that contains all the results.

All numeric values are rounded to three significant digits.

Explanation of columns:

* gene - `r assembly` gene measured
* group 1 - control group
* group 2 - treatment group
* baseMean - mean across samples
* padj - false discovery rate; adjusted p-value computed using method of Benjamini and Hochberg
* log2FoldChange - log2(group 2 average/group 1 average)
* lfcsE - log2FoldChange standard error
* stat - test statistic used to assess significance 
* description - gene description 
* SL4 - putative SL4 (June 2019 assembly release) gene counterpart

* * *

# Discussion

Within each genotype, the number of genes exhibiting expression changes increased with treatment duration. 

The different genotypes exhibited different numbers of differentially expressed genes, with the `are` genotype exhibiting the greatest number. This is consistent with the known `are` lower fertility mutant phenotype.

* * *

# Conclusion

* The number of genes detected as changed increased with treatment duration.

* Most fold-changes for genes that changed in respond to the treatment were positive, indicating that expression levels increased.

* We observed more temperature-dependent gene expression differences in the `are` genotype than for the other two genotypes tested.

* * *

# Session info

```{r}
sessionInfo()
```